python/pyarrow/tests/test_dataset.py

```diff
@@ -4912,3 +4912,33 @@ def test_read_table_nested_columns(tempdir, format):
         {'user_id': 'qrs456', 'type': 'scroll', 'values': [None, 3, 4],
          'structs': [{'fizz': 'buzz', 'foo': None}], 'a.dotted.field': 2}
     ]
+
+
+def test_dataset_partition_with_slash(tmpdir):
+    from pyarrow import dataset as ds
+
+    path = tmpdir / "slash-writer-x"
+
+    dt_table = pa.Table.from_arrays([
+        pa.array([1, 2, 3, 4, 5], pa.int32()),
+        pa.array(["experiment/A/f.csv", "experiment/B/f.csv",
+                  "experiment/A/f.csv", "experiment/C/k.csv",
+                  "experiment/M/i.csv"], pa.utf8())], ["exp_id", "exp_meta"])
+
+    ds.write_dataset(
+        data=dt_table,
+        base_dir=path,
+        format='parquet',
+        partitioning=['exp_meta'],
+        partitioning_flavor='hive',
+    )
+
+    read_table = ds.dataset(
+        source=path,
+        format='parquet',
+        partitioning='hive',
+        schema=pa.schema([pa.field("exp_id", pa.int32()),
+                         pa.field("exp_meta", pa.utf8())])
+    ).to_table().combine_chunks()
+
```

Comment:
Could we also assert what the escaped partition directories are named? Given we are trying to be compatible with other systems, it seems like it would be wise to enforce what the Uri-encoded form is, rather than just asserting we can roundtrip.

Reply:
Shouldn’t it be handled by the encoder test cases?

Reply:
That tests that the encoding function works, but not that it is used to write the directory names.

I was thinking of a test like this:

```python
   file_paths = list(path.iterdir())
   assert Path("exp_meta=experiment%2FA%2Ff.csv/part-0.parquet") in file_paths
```

Reply:
I see your point. We could write a test, then it should reflect platform dependent test cases too?

Reply:
I added a check.